src/stream.rs

```diff
@@ -0,0 +1,229 @@
+extern "C" fn write_callback<W: io::Write + 'static>(
+    env: *mut c_void,
+    data: *mut c_uchar,
+    length: c_uint,
+) -> cairo_status_t {
+    // This is consistent with the type of `env` in `Surface::_for_stream`.
+    let env: *const UnsafeCell<CallbackEnvironment> = env as _;
+
+    // Safety: the user data entry keeps `Rc<UnsafeCell<CallbackEnvironment>>` alive
+    // until the surface is destroyed.
+    // If this is called by cairo, the surface is still alive.
+    let env: &UnsafeCell<CallbackEnvironment> = unsafe { &*env };
+
+    with_mut_env(env, |env| {
+        if let Some(stream) = &mut env.stream {
+            // Safety: `write_callback<W>` was instanciated in `Surface::_for_stream`
+            // with a W parameter consistent with the box that was unsized to `Box<dyn Any>`.
+            let stream = unsafe {
+                stream.downcast_mut_unchecked::<W>()
+            };
+            // Safety: this is the callback contract from cairo’s API
+            let data = unsafe {
+                std::slice::from_raw_parts(data, length as usize)
+            };
+            // Because `<W as Write>::write_all` is a generic,
+            // we must conservatively assume that it can panic.
+            let result = std::panic::catch_unwind(AssertUnwindSafe(|| stream.write_all(data)));
+            match result {
+                Ok(Ok(())) => {
+                    return Status::Success
+                }
+                Ok(Err(error)) => {
+                    env.io_error = Some(error);
+                }
+                Err(payload) => {
+                    env.unwind_payload = Some(payload);
+                }
+            }
+        }
+        Status::WriteError
+    }).into()
+}
+
+fn with_mut_env<R, F>(env: &UnsafeCell<CallbackEnvironment>, f: F) -> R
+    where F: FnOnce(&mut CallbackEnvironment) -> R
+{
+    // Safety:
+    //
+    // * The only long-lived pointers to this environment are
+    //   the `void` pointer passed to surface constructor, and the one in user data.
+    // * `STREAM_CALLBACK_ENVIRONMENT` is private,
+    //   so the user data entry is only accessible in this module.
+    // * `Surface` is !Send and !Sync, so a given surface is only used on a single thread.
+    //
+    // Therefore, there are no other ongoing reference to (part of) this `CallbackEnvironment`
+    // and it is sound to claim exclusive / mutable access with `&mut`.
+    unsafe {
+        f(&mut *env.get())
+    }
+}
+
+struct RawStream<W>(*mut W);
+
+impl<W: io::Write> io::Write for RawStream<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> { unsafe { (*self.0).write(buf) }}
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> { unsafe { (*self.0).write_all(buf) }}
+    fn flush(&mut self) -> io::Result<()> { unsafe { (*self.0).flush() } }
+}
+
+trait AnyExt {
+    /// Any::downcast_mut, but YOLO
+    unsafe fn downcast_mut_unchecked<T>(&mut self) -> &mut T {
```

Comment:
Does it make any measurable performance difference compared to the safe version?

Reply:
I haven’t measured it, but I’m relatively confident about this particular aspect of the unsafety of the implementation. See the comment at the call site: https://github.com/gtk-rs/cairo/pull/260/files/cf2f6925ffe09f45429afd1940731b57454c6d49#diff-2763e33724cc9b239c750ae973480953R168

I did encounter segfauts initially when this method was a free function taking a `&mut dyn Any` parameter. I would pass it `&mut Box<dyn Any>` expecting auto-deref, but the unsizing coercion of `Box<dyn Any>` to `dyn Any` (adding another layer of trait object indirection) took precedence. So I would end up with a reference at the wrong address, even though `debug_assert!(stream.is::<W>());` passed! Making this a method in an extension trait fixes this because unsizing coercion desn’t happen for method receivers like it does for call arguments.

Reply:
Ok, it *seems* fine to me too so let's keep it then :) Thanks for the details